server/src/main/scala/com/cloudera/livy/server/ApiVersioningSupport.scala

```diff
@@ -0,0 +1,98 @@
+package com.cloudera.livy.server
+
+import javax.servlet.http.HttpServletRequest
+
+import org.scalatra.{NotAcceptable, ScalatraBase}
+
+/**
+  * Livy's servlets can mix-in this trait to get API version support.
+  *
+  * Example: {{{
+  * import ApiVersions._
+  * class FooServlet
+  *   ...
+  *   with ApiVersioningSupport
+  *   ...
+  * {
+  *   get("/test") {
+  *     ...
+  *   }
+  *   get("/test", getApiVersion() <= v0_2) {
+  *     ...
+  *   }
+  *   get("/test", getApiVersion() <= v0_1) {
+  *     ...
+  *   }
+  * }
+  * }}}
+  */
+trait ApiVersioningSupport extends AbstractApiVersioningSupport {
+  this: ScalatraBase =>
+  // Link the abstract trait to Livy's version enum.
+  override val apiVersions = ApiVersions
+  override type ApiVersionType = ApiVersions.Value
+}
+
+trait AbstractApiVersioningSupport {
+  this: ScalatraBase =>
+  protected val apiVersions: Enumeration
+  protected type ApiVersionType
+
+  /**
+    * Before proceeding with routing, validate the specified API version in the request.
+    * If validation passes, cache the parsed API version as a per-request attribute.
+    */
+  before() {
+    try {
+      request(AbstractApiVersioningSupport.ApiVersionKey) = parseForApiVersion(request)
+    } catch {
+      case e: Exception =>
+        halt(NotAcceptable(e.getMessage))
+    }
+  }
+
+  /**
+    * @return The specified API version in the request.
+    */
+  def getApiVersion(): ApiVersionType = {
+    request(AbstractApiVersioningSupport.ApiVersionKey).asInstanceOf[ApiVersionType]
+  }
+
+  private def extractVersionFromAcceptHeader(acceptHeader: String): Option[String] = {
+    // Get every character after "application/vnd.livy.v" until hitting a + sign.
+    val acceptHeaderRegex = """application/vnd\.livy\.v([^\+]*).*""".r
+
+    acceptHeader match {
+      case acceptHeaderRegex(v) => Option(v)
+      case _ => None
+    }
+  }
+
+  private def parseForApiVersion(request: HttpServletRequest): ApiVersionType = {
```

Comment:
It took me a while to follow this code and understand what it's doing in which cases... I'd like to suggest a shorter and (to me) clearer version:

    private val acceptHeaderRegex = """application/vnd\.livy\.v([^\+]*).*""".r
    
    private def parseApiVersion(request: HttpServletRequest): Unit = {
      val version = header("Accept") match {
        case Some(accept @ acceptHeaderRegex(v)) =>
          try {
            apiVersions.withName(v).asInstanceOf[ApiVersionType]
          } catch {
            case e: Exception =>
              halt(s"Unknown API version: $accept")
          }
    
        case _ =>
          apiVersions.apply(apiVersions.maxId - 1).asInstanceOf[ApiVersionType]
      }
      request(AbstractApiVersioningSupport.ApiVersionKey) = version
    }

This may or may not compile - I did some basic testing on the Scala shell. But mainly, it would be nice to have a comment saying what's going on; basically, no Accept header means latest version, and so does an accept header that does not have a Livy version in it.